packages/testing-lib/src/context.ts

```diff
@@ -0,0 +1,129 @@
+ * context. It should not be used after the context has exited and doing
+ * so should be treated as undefined behavior.
+ *
+ * The `within` function will:
+ *  - call the `enter` method exactly once.
+ *  - call the `exit` exactly once and only after `enter` has been called.
+ */
+export interface Context<THandle> {
+  [enter]: MaybeAsyncFunction<[], THandle>;
+  [exit]?: MaybeAsyncFunction<[THandle], void>;
+}
+
+type HandleFromEnter<
+  TEnter extends MaybeAsyncFunction<[], any>
+> = TEnter extends MaybeAsyncFunction<[], infer THandle> ? THandle : never;
+
+interface BuildContext<
+  TEnter extends MaybeAsyncFunction<[], any>,
+  TExit extends MaybeAsyncFunction<[HandleFromEnter<TEnter>], void>
+> {
+  [enter]: TEnter;
+  [exit]?: TExit;
+}
+
+// Can't deep resolve since TypeScript doesn't allow recursive type constructors
+// Replace this with the upcoming `awaited T` type when it's released.
+type ResolveShallow<T> = T extends PromiseLike<infer U> ? U : T;
+
+/**
+ * This helper ensures that the return value of `within` is typed
+ * as `PromiseLike<TReturn>` since TypeScript doesn't know that
+ * `PromiseLike<PromiseLike<TReturn>>` is impossible.
+ */
+type ContextResult<
+  TContext extends Context<any>,
+  TReturn
+> = TContext[typeof enter] extends (...args: Array<any>) => PromiseLike<any>
+  ? PromiseLike<ResolveShallow<TReturn>>
+  : TContext[typeof exit] extends (...args: Array<any>) => PromiseLike<any>
+  ? PromiseLike<ResolveShallow<TReturn>>
+  : TReturn;
+
+type Result<T, E = unknown> = { value: T } | { error: E };
+
+function handleResult<T>(result: Result<T>): T {
+  if ('value' in result) {
+    return result.value;
+  }
+  throw result.error;
+}
+
+function isPromiseLike<T>(obj: PromiseLike<T> | T): obj is PromiseLike<T> {
+  return (
+    (typeof obj === 'object' || typeof obj === 'function') &&
+    'then' in obj &&
+    typeof obj.then === 'function'
+  );
+}
+
+// This seems too sophisticated for TypeScript to figure out.
+function castReturn<TContext extends Context<any>, TReturn>(
+  value: TReturn | PromiseLike<TReturn>,
+): ContextResult<TContext, TReturn> {
+  return value as ContextResult<TContext, TReturn>;
+}
+
+export function within<
+  // This is generic with respect to the enter method so that it becomes the
+  // source of truth for the handle's type. Type error(s) will occur on the
+  // exit method or provided function if they don't have the correct handle type.
+  TEnter extends MaybeAsyncFunction<[], any>,
+  TExit extends MaybeAsyncFunction<[HandleFromEnter<TEnter>], void>,
+  TReturn
+>(
+  context: BuildContext<TEnter, TExit>,
+  fn: MaybeAsyncFunction<[HandleFromEnter<TEnter>], TReturn>,
+): ContextResult<BuildContext<TEnter, TExit>, TReturn> {
+  const handle = context[enter]();
+
+  if (isPromiseLike(handle)) {
+    return castReturn(handle.then(performFn));
+  }
+
+  return castReturn(performFn(handle));
+
+  function performFn(handle: HandleFromEnter<TEnter>) {
```

Comment:
Dirty hoisting! I always find this code structure case interesting because it serves different types of reading code. I'm not going to take a stance on this for now.